Source/ARTRest.m

```diff
@@ -788,5 +788,12 @@ - (void)setAndPersistAPNSDeviceToken:(NSString *)deviceToken {
     });
 }
 
+- (void)resetDevice {
+    dispatch_sync(ARTRestInternal.storageQueue, ^{
+        ARTLocalDevice *device = [self device_nosync];
```

Comment:
My understanding of the `*_noSync` methods is that they need to be manually synchronised _with the rest of the library_ – i.e. they need to be called on the `clientOptions`’ `internalDispatchQueue`. Although you're calling `device_noSync` on _a_ queue here, I don't _think_ that's enough.

Reply:
Originally the caller in `ARTPushActivationStateWaitingForDeregistration` did this:

```
ARTLocalDevice *local = self.machine.rest.device_nosync;
[local setAndPersistIdentityTokenDetails:nil];
```

So I thought that 

```
- (void)resetDevice {
    dispatch_sync(ARTRestInternal.storageQueue, ^{
        ARTLocalDevice *device = [self device]; // <-- ARTRest queue sync call
```

Would be too much. I'm still not sure about this though.

Reply:
The difference is that the original `ARTPushActivationStateWaitingForDeregistration` code 
[was always called on the `ARTRest` internal queue](https://github.com/ably/ably-cocoa/blob/3b6e33508450395c0b64357632a2897a4c585d06/Source/ARTPushActivationStateMachine.m#L88-L92) and hence could use the `*_noSync` methods. So I think you need to use the `device` method here — why do you think it might be "too much"?